src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,143 @@
+            throw new IndexOutOfBoundsException("Invalid index: " + index + " for size: " + size);
+        }
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            tail = new Node(null, value, tail);
+        } else {
+            Node nextNode = getNode(index);
+            nextNode.prev.next = new Node(nextNode.prev, value, nextNode);
+            nextNode.prev = nextNode.prev.next;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexValidation(index);
+        return getNode(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexValidation(index);
+        T oldValue = get(index);
+        getNode(index).element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexValidation(index);
+        Node node = getNode(index);
+        T removedElement = get(index);
+        if (node == tail) {
+            if (size == 1) {
+                tail = null;
+            } else {
+                tail = node.next;
+                tail.prev = null;
+            }
+        } else if (node == head) {
+            head = node.prev;
+            head.next = null;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+        size--;
+        return removedElement;
     }
 
     @Override
     public boolean remove(T object) {
+        Node node = tail;
+        for (int i = 0; node != null; i++) {
+            if (node.element == object
+                    || node.element != null && node.element.equals(object)) {
+                remove(i);
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
+
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndexValidation(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index + " for size: " + size);
+        }
+    }
+
+    private Node getNode(int index) {
+        if (index == 0) {
+            return tail;
+        }
+        if (index == size - 1) {
+            return head;
+        }
+        if (index < size / 2) {
+            Node node = tail;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+            return node;
+        }
+        Node node = head;
+        for (int i = size - 1; i > index; i--) {
+            node = node.prev;
+        }
+        return node;
+    }
+
+    private class Node {
```

Comment:
```suggestion
    private class Node<T> {
```

Reply:
> 

For what we need parametrize class if we can just make it non-static and use Type from LinkedList It will be the same ? I think it is useless action